alibi/explainers/similarity/grad.py

```diff
@@ -0,0 +1,203 @@
+        Returns
+        -------
+        X:
+            Input data formatted for explain method.
+        Y:
+            Target data formatted for explain method.
+
+        """
+        X = self._match_shape_to_data(X, 'x')
+        if isinstance(X, np.ndarray):
+            X = self.backend.to_tensor(X)
+
+        if self.task == 'regression' and Y is None:
+            err_msg = 'Regression task requires a target value. `Y` must be provided, either as a value or a function.'
+            raise ValueError(err_msg)
+
+        if Y is None:
+            Y = self.predictor(X)
+            Y = self.backend.argmax(Y)
+        elif callable(Y):
+            Y = Y(X)
+
+        Y = self._match_shape_to_data(Y, 'y')
+        if isinstance(Y, np.ndarray):
+            Y = self.backend.to_tensor(Y)
+
+        return X, Y
+
+    def explain(
+            self,
+            X: 'Union[np.ndarray, tensorflow.Tensor, torch.Tensor]',
+            Y: 'Optional[Union[np.ndarray, tensorflow.Tensor, torch.Tensor, Callable]]' = None) -> "Explanation":
+        """Explain the predictor's predictions for a given input.
+
+        Computes the similarity score between the input and the training set. Reorders the training set according to the
+        score in descending order. Returns an explainer object containing the scores and the corresponding training set
+        instances as well as the most and least similar instances of the data set.
+
+
+        Parameters
+        ----------
+        X:
+            `X` can be a `numpy` array, `tensorflow` tensor, or `torch` tensor of same shape as the training data with
+            or without the batch dimension. If the batch dimension is missing it's added.
+        Y:
+            `Y` can be a `numpy` array, `tensorflow` tensor, `torch` tensor or a function that returns one of these. It
+            must either be or return a value of the same shape as `X`. If the batch dimension is missing it's added. In
+            the case of a regression task the `Y` argument must be present. If the task is classification then `Y`
+            defaults to the model prediction.
+
+        Returns
+        -------
+        `Explanation` object containing the ordered similarity scores for the instance with additional metadata as \
+        attributes. Contains the following data-related attributes
+            -  `scores`: ``np.array`` - similarity scores for each instance in the training set.
+            -  `X_train`: ``np.array`` - training set instances in the order of descending similarity scores.
+            -  `Y_train`: ``np.array`` - training set labels in the order of descending similarity scores.
+            -  `most_similar`: ``np.array`` - most similar instances to the input.
+            -  `least_similar`: ``np.array`` - least similar instances to the input.
+
+        Raises
+        -------
+        ValueError:
+            If `Y` is `None` and the `task` is `regression`.
+        """
+        self._verify_fit()
```

Comment:
I think something is missing. First, I don't see a `fit` method for the public class. Second, it these private methods don't seem to exist, are they supposed to be called on the backends?

Reply:
Ignore the second comment, I didn't consider the inheritance. However, `fit` should be a real method on the public interface for documentation purposes.